### PR TITLE
Balancing the subscribers

### PR DESCRIPTION
**Pull request by dennisdoomen**
Introduced a page request queue so that a fast subscriber does not prevent slow subscribers from receiving pages of transactions. Also adds a lot of new tests to verify more edge cases and interoperability between pre-loading and caching.

### CODE REVIEW

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 1]
```diff
@@ -81,50 +106,99 @@ public IDisposable Subscribe(long? lastProcessedCheckpoint, Subscriber subscribe
...
+            
+            pendingRequests.Enqueue(loadRequest);
+            requestQueued.TrySetResult(true);
+
+            return loadRequest.Page;
+        }
+
+        private bool ShouldPreloadNextPage(Page completedPage)
+        {
+            // If there's no cache configured, preloading will only benefit if the subscriber requests the next page while it is still in the queue.
```
Comment (ronaldkroon) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: I can't relate this comment to the code below.

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 2]
```diff
@@ -134,276 +208,190 @@ private Page TryGetNextPageFromCache(long precedingCheckpoint, string subscripti
                     {
                         resultPage.Add(cachedTransaction);
```
Comment (ronaldkroon) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: Looks like we're adding the same cachedTransaction every time. Why?

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs (hunk 1 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: Tried to rephrase it

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 3]
```diff
@@ -81,50 +106,101 @@ public IDisposable Subscribe(long? lastProcessedCheckpoint, Subscriber subscribe
...
+            CancellationToken cancellationToken)
+        {
+            if (disposingCancellationTokenSource.IsCancellationRequested)
             {
                 throw new ObjectDisposedException(typeof(PollingEventStoreAdapter).FullName);
             }
 
-            Page pageFromCache = TryGetNextPageFromCache(lastProcessedCheckpoint, subscriptionId);
+            Page pageFromCache = TryGetNextPageFromCache(precedingCheckpoint, subscriptionId);
             if (pageFromCache.Transactions.Count > 0)
```
Comment (corvansteijn) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: :wrench: `if (pageFromCache.IsNotEmpty)` ?

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 4]
```diff
@@ -81,50 +106,101 @@ public IDisposable Subscribe(long? lastProcessedCheckpoint, Subscriber subscribe
...
+            lock (subscriptionLock)
+            {
+                subscriptions.Remove(subscription);
+            }
+        }
+
+        internal async Task<Page> GetNextPage(long precedingCheckpoint, string subscriptionId,
+            CancellationToken cancellationToken)
+        {
+            if (disposingCancellationTokenSource.IsCancellationRequested)
```
Comment (corvansteijn) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: 🔧 Cancellation could be requested right after this check, but before the actual implementation of the method. That means you will always have to be prepared for the code to fail because of unexpected disposing. Why even guard against it here?

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 5]
```diff
@@ -81,50 +106,101 @@ public IDisposable Subscribe(long? lastProcessedCheckpoint, Subscriber subscribe
...
             }
 
-            Page pageFromCache = TryGetNextPageFromCache(lastProcessedCheckpoint, subscriptionId);
+            Page pageFromCache = TryGetNextPageFromCache(precedingCheckpoint, subscriptionId);
             if (pageFromCache.Transactions.Count > 0)
             {
                 return pageFromCache;
             }
 
+            LoadRequest pendingRequest =
```
Comment (corvansteijn) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs:151]: 🔧 So we will request to load the page, even when there is already a request for `precedingCheckpoint - 1`. Although there can be gaps in the checkpoint numbers, we do know that most of the page will be loaded for all request where `precedingCheckpoint - pageSize * someFillFactor`.

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 6]
```diff
@@ -81,50 +106,101 @@ public IDisposable Subscribe(long? lastProcessedCheckpoint, Subscriber subscribe
...
+                CancellationToken = cancellationToken
+            };
+            
+            pendingRequests.Enqueue(loadRequest);
+            requestQueued.TrySetResult(true);
+
+            return loadRequest.Page;
+        }
+
+        private bool ShouldPreloadNextPage(Page completedPage)
```
Comment (corvansteijn) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: ❓ Could it be that (most of) the next page is already in the cache or was already requested? 
🔧 In that case, we should not request it.

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 7]
```diff
@@ -134,276 +210,190 @@ private Page TryGetNextPageFromCache(long precedingCheckpoint, string subscripti
...
             return new Page(precedingCheckpoint, transactions);
         }
 
+        private void CachePage(Page page, string subscriptionId)
+        {
+            IReadOnlyList<Transaction> transactions = page.Transactions;
+
+            if (cachedTransactionsByPrecedingCheckpoint != null)
+            {
+                /* Add to cache in reverse order to prevent other projectors
```
Comment (corvansteijn) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: 😕 It took me quite a while before I realized that this would prevent other projectors to request already loaded transactions because the request stays on the queue until this method is done and that projectors will first check if there is already a request on the queue and only then check the cache.

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 8]
```diff
@@ -81,50 +106,101 @@ public IDisposable Subscribe(long? lastProcessedCheckpoint, Subscriber subscribe
...
+            return completedPage;
+        }
+
+        private Task<Page> RequestPageload(string subscriptionId, long precedingCheckpoint, CancellationToken cancellationToken)
+        {
+            var loadRequest = new LoadRequest
+            {
+                SubscriptionId = subscriptionId,
+                PrecedingCheckpoint = precedingCheckpoint,
+                CancellationToken = cancellationToken
```
Comment (corvansteijn) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: 🔧 The `CancellationToken` property is never used

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 9]
```diff
@@ -81,50 +106,101 @@ public IDisposable Subscribe(long? lastProcessedCheckpoint, Subscriber subscribe
...
         private Page TryGetNextPageFromCache(long precedingCheckpoint, string subscriptionId)
         {
-            if ((cachedTransactionsByPrecedingCheckpoint != null) && cachedTransactionsByPrecedingCheckpoint.TryGet(precedingCheckpoint, out Transaction cachedTransaction))
+            if ((cachedTransactionsByPrecedingCheckpoint != null) &&
+                cachedTransactionsByPrecedingCheckpoint.TryGet(precedingCheckpoint, out Transaction cachedTransaction))
             {
-                var resultPage = new List<Transaction>(maxPageSize) { cachedTransaction };
+                var resultPage = new List<Transaction>(maxPageSize) {cachedTransaction};
 
                 while (resultPage.Count < maxPageSize)
```
Comment (corvansteijn) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs:186]: ❌ As discussed. If only the first N transactions of the requested page are in the cache, this will loop forever.

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 10]
```diff
@@ -134,276 +210,190 @@ private Page TryGetNextPageFromCache(long precedingCheckpoint, string subscripti
...
+            lock (subscriptionLock)
             {
-                if (!isDisposed)
+                if (!disposingCancellationTokenSource.IsCancellationRequested)
                 {
-                    isDisposed = true;
+                    disposingCancellationTokenSource.Cancel();
 
-                    foreach (Subscription subscription in Subscriptions.ToArray())
+                    foreach (Subscription subscription in subscriptions.ToArray())
```
Comment (corvansteijn) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: 🔧 Since you already lock the subscriptions hashset when modifying it, you probably don't need the ToArray. (Or could you maybe remove the lock?)

**Review by corvansteijn — CHANGES_REQUESTED** [review]

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs (hunk 5 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs:151]: Caching will take care of that. By the time that previous request is done, most of it will be in the cache.

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs (hunk 6 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: True, but by the time the actual request comes in from the subscriber, it may have been vacated from the cache. But maybe we can check the cache to see if a majority (TBD) is in the cache already.

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs (hunk 5 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs:151]: #16

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs (hunk 6 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: #16

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs (hunk 7 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: #16

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs (hunk 9 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs:186]: Already fixed

**Comment by dennisdoomen** [comment]
Fixed all comments. Added a separate issue #16 for optimizing the usage of the cache.

**Comment by dennisdoomen** [comment]
> Fixed all comments. Added a separate issue #16 for optimizing the usage of the cache.

@corvansteijn @ronaldkroon anything else?

**Review by corvansteijn — APPROVED** [review]

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 11]
```diff
@@ -81,50 +106,83 @@ public IDisposable Subscribe(long? lastProcessedCheckpoint, Subscriber subscribe
...
-                StartPreloadingNextPage(loadedPage.LastCheckpoint, subscriptionId);
+                logDebug(() =>
+                    $"Pending Request Found (subscription: {subscriptionId}, originalSubscription: " +
+                    $"{pendingRequest.SubscriptionId}, preceding checkpoint: {precedingCheckpoint}");
+
+                return await pendingRequest.Page;
             }
 
-            return loadedPage;
+            var completedPage = await RequestPageload(subscriptionId, precedingCheckpoint, cancellationToken);
```
Comment (ronaldkroon) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: 🙃 use explicit type.

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 12]
```diff
@@ -136,274 +194,212 @@ private Page TryGetNextPageFromCache(long precedingCheckpoint, string subscripti
...
-                $"Subscription {subscriptionId} has not found the next transaction in the cache.");
-#endif
+            logDebug(() =>
+                $"Cache Miss: (subscription: {subscriptionId}, precedingCheckpoint: {precedingCheckpoint}).");
 
             return new Page(precedingCheckpoint, new Transaction[0]);
         }
 
-        private void StartPreloadingNextPage(long precedingCheckpoint, string subscriptionId)
+        private Task<Page> RequestPageload(string subscriptionId, long precedingCheckpoint, CancellationToken cancellationToken)
```
Comment (ronaldkroon) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: 🙃 RequestPage_L_oad

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 13]
```diff
@@ -81,50 +106,83 @@ public IDisposable Subscribe(long? lastProcessedCheckpoint, Subscriber subscribe
...
+            }
+
+            return completedPage;
+        }
+
+        private bool ShouldPreloadNextPage(Page completedPage)
+        {
+            // NOTE: Only preload when caching is enabled. If it is not, other subscribers will only benefit from preloading if the request for a particular
+            // page is still in the queue.
+            return (cachedTransactionsByPrecedingCheckpoint != null) && (completedPage?.Transactions.Count == maxPageSize);
```
Comment (ronaldkroon) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: 🔧 Replace left condition with property named IsCachingEnabled.

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 14]
```diff
@@ -81,50 +106,83 @@ public IDisposable Subscribe(long? lastProcessedCheckpoint, Subscriber subscribe
...
+
+                return await pendingRequest.Page;
             }
 
-            return loadedPage;
+            var completedPage = await RequestPageload(subscriptionId, precedingCheckpoint, cancellationToken);
+            if (ShouldPreloadNextPage(completedPage))
+            {
+                // Ingore the result. It will become available through the queue.
+                Task _ = RequestPageload(subscriptionId, completedPage.PrecedingCheckpoint, cancellationToken);
```
Comment (ronaldkroon) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: 🤔 Shouldn't we request a page based on the last checkpoint of the completed page?

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs (hunk 13 above)
Comment (ronaldkroon) [reply Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: ❓ Could the `completePage` ever be `null`?

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 15]
```diff
@@ -136,274 +194,212 @@ private Page TryGetNextPageFromCache(long precedingCheckpoint, string subscripti
...
+    internal class LoadRequest
+    {
+        private readonly TaskCompletionSource<Page> completionSource;
+
+        public LoadRequest()
+        {
+            completionSource = new TaskCompletionSource<Page>();
+        }
+
+        public string SubscriptionId { get; set; }
```
Comment (ronaldkroon) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs:393]: 🙃 Why are all the setters public?

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 16]
```diff
@@ -136,274 +194,212 @@ private Page TryGetNextPageFromCache(long precedingCheckpoint, string subscripti
...
-            TaskCompletionSource<Page> loaderCompletionSource, string subscriptionId)
+        private async Task WaitForPendingRequest()
         {
-            Page nextPage;
-
-            try
+            if (pendingRequests.IsEmpty)
             {
-                try
+                await requestQueued.Task;
```
Comment (ronaldkroon) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: ❓ Why don't we use the cancellation token here?

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 17]
```diff
@@ -136,274 +194,212 @@ private Page TryGetNextPageFromCache(long precedingCheckpoint, string subscripti
...
-                if (candidatePage.PrecedingCheckpoint == precedingCheckpoint)
-                {
-                    return candidatePage;
-                }
-            }
+            return loadRequest.Page;
         }
 
-        private Task<Page> TryLoadNextOrWaitForPreLoadingToFinish(long precedingCheckpoint, string subscriptionId)
+        private void RunAsync()
```
Comment (ronaldkroon) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: 🔧 This is not a very descriptive method name.

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs [hunk 18]
```diff
@@ -136,274 +194,212 @@ private Page TryGetNextPageFromCache(long precedingCheckpoint, string subscripti
...
+                logDebug(() =>
+                    $"No Transactions: (subscriptionId: {subscriptionId}, precedingCheckpoint: {precedingCheckpoint})");
             }
 
             return new Page(precedingCheckpoint, transactions);
         }
 
+        private void CachePage(Page page, string subscriptionId)
+        {
+            IReadOnlyList<Transaction> transactions = page.Transactions;
```
Comment (ronaldkroon) [inline Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: 🙃 Move into `if`.

**Review by ronaldkroon — CHANGES_REQUESTED** [review]

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs (hunk 13 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs]: > Could the completePage ever be null?

Yes, it can. Made that a bit more explicit.

File: Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs (hunk 15 above)
Comment (dennisdoomen) [reply Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs:393]: The class is `internal`

**Comment by dennisdoomen** [comment]
All fixes applied
